[inductor] Respect dtype argument in ops.constant

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92093

Consider the following example:

```python
def fn(x):
    y = torch.full_like(x, 1.2, dtype=torch.int64)
    return x + y
```

In eager this truncates 1.2 to 1, then adds it to `x`. However, in
inductor the literal "1.2" is used verbatim and the result is off by
0.2. This fixes the issue by respecting the dtype argument to `ops.constant`
and truncating accordingly.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
[inductor] Respect dtype argument in ops.constant

Consider the following example:

```python
def fn(x):
    y = torch.full_like(x, 1.2, dtype=torch.int64)
    return x + y
```

In eager this truncates 1.2 to 1, then adds it to `x`. However, in
inductor the literal "1.2" is used verbatim and the result is off by
0.2. This fixes the issue by respecting the dtype argument to `ops.constant`
and truncating accordingly.

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -142,7 +142,8 @@ def to_dtype(x, dtype: torch.dtype):
     def constant(value, dtype):
-        return triton_constant(value)
+        type_ = torch._prims_common.dtype_to_type(dtype)
+        return triton_constant(type_(value))
```
This still doesn't respect the size of `dtype`, e.g. 300 could pass through `torch.int8` without being truncated. However, that would be undefined behavior in C++ anyway so I'm not sure it matters. Though this may be worth thinking about for  `uint8`.

lezcano replied:
I'd say that it'd be good to have a check as it may uncover issues. Worst comes worst, we could put it behind a debug flag.

lezcano reviewed (approved):
Nice! Let's wait for approval from someone from Meta, but otherwise let's merge this one first and I'll update my PR afterwards accordingly.

peterbell10 committed:
Update on "[inductor] Respect dtype argument in ops.constant"


Consider the following example:

```python
def fn(x):
    y = torch.full_like(x, 1.2, dtype=torch.int64)
    return x + y
```

In eager this truncates 1.2 to 1, then adds it to `x`. However, in
inductor the literal "1.2" is used verbatim and the result is off by
0.2. This fixes the issue by respecting the dtype argument to `ops.constant`
and truncating accordingly.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Respect dtype argument in ops.constant"


Consider the following example:

```python
def fn(x):
    y = torch.full_like(x, 1.2, dtype=torch.int64)
    return x + y
```

In eager this truncates 1.2 to 1, then adds it to `x`. However, in
inductor the literal "1.2" is used verbatim and the result is off by
0.2. This fixes the issue by respecting the dtype argument to `ops.constant`
and truncating accordingly.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Lint on "[inductor] Respect dtype argument in ops.constant"


Consider the following example:

```python
def fn(x):
    y = torch.full_like(x, 1.2, dtype=torch.int64)
    return x + y
```

In eager this truncates 1.2 to 1, then adds it to `x`. However, in
inductor the literal "1.2" is used verbatim and the result is off by
0.2. This fixes the issue by respecting the dtype argument to `ops.constant`
and truncating accordingly.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge -f "ASAN timeout is not related"
